DOC: add details on default build options used by meson-python (#338)
mesonbuild/meson-python: docs/explanations/default-options.rst

.. SPDX-FileCopyrightText: 2023 The meson-python developers
..
.. SPDX-License-Identifier: MIT

.. _explanations-default-options:

*********************
Default build options
*********************

Meson offers many `built-in options <https://mesonbuild.com/Builtin-options.html>`__,
and in the vast majority of cases those have good defaults. There are a couple

> of cases however where ``meson-python`` either is forced or chooses to override

Reviewer: Forced?

Author: I'll rephrase as "needs to" if that's better. Overrides for, e.g., `--prefix` are necessary.

Reviewer: Technically, they are not necessary. We could very well use the default prefix and use the `$DESTDIR` environment variable to install in a temporary directory. I would even argue that this would be better. Anyhow, it does not matter.

Reviewer 2: Yes, I think it was only needed for the Debian heuristics, I think we can drop it now that Meson uses directory placeholders for all paths.

those with its own defaults. To view what those are for the version of
``meson-python`` you have installed, look at the *User defined options* section
of the output during the configure stage of the build (e.g., by running
``python -m build --wheel``). This will look something like:

.. code-block:: text

    User defined options
      Native files     : /home/username/code/project/.mesonpy-native-file.ini
      debug            : false
      optimization     : 2
      prefix           : /home/username/mambaforge/envs/project-dev
      python.platlibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      python.purelibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      b_ndebug         : if-release

Let's go through each option and why they are used:

> - meson-python uses a native file, named ``.mesonpy-native-file.ini`` in order

Reviewer: The native file will not named like this anymore in the next meson-python release. It will be in the build directory and names `meson-python-native-file.ini`. One thing that is not obvious is that multiple `--native-file` options can be passed and Meson will merge the contents. Therefore, if the user needs to pass any other configuration via a native file, they can add another `--native-file` argument, as long as they do not try to set the path for the `python` executable. I think it would be useful to mention this in the documentation.

Author: Agreed, thanks. Will update and add that info.

Author: added now

  to point Meson at the correct ``python`` interpreter to use (the same one for
  which ``meson-python`` was installed). This is necessary, because Meson may
  otherwise look for the first Python interpreter on the PATH (usually the same
  one, but not always the case).
- The ``prefix`` and ``platlibdir``/``purelibdir`` options also point Meson at

>   that same interpreter and the environment in which it is installed.

Reviewer: These are going away, see #279.

Author: That's merged, but these options are still there. They're not specific to editable installs, so I'm not sure I understand what you meant here.

Reviewer 2: I think Daniele meant #340. I'll review that tomorrow, but we can either wait until that gets a resolution, or merge this and have that PR update it.

Reviewer: Yes, sorry, I linked to the wrong PR.

Reviewer 2: Because of #341 and #344, we cannot get rid of the these options at the moment, so they should stay for now.

Author: The current text is good as is given the current state of the repo, I think we can resolve this comment.

.. SPDX-FileCopyrightText: 2023 The meson-python developers
..
.. SPDX-License-Identifier: MIT

.. _explanations-default-options:

*********************
Default build options
*********************

Meson offers many `built-in options <https://mesonbuild.com/Builtin-options.html>`__,
and in the vast majority of cases those have good defaults. There are a couple
of cases however where ``meson-python`` either needs to or chooses to override

> those with its own defaults. To view what those are for the version of

Reviewer 2: possible improvement:

We could split the paragraph here and also document why we might choose to override the defaults. I wrote about the reasoning in https://github.com/mesonbuild/meson-python/pull/325#issuecomment-1463902837.

We could also note that the options we do need to set are mostly for build/installation details that should not be relevant to the user. The only exception is the native file, but the user should be able to override the parts of it that make sense (IIRC everything but the Python interpreter).

Reviewer: If we keep `meson install` I would much prefer if we would get rid of the `--prefix` and `--python.{plat,pure}lib` options and use the `$DESTDIR` environment variable pointing to somewhere in `/tmp` instead. This would have the additional benefit of putting the installation directory on tmpfs, with possible significant speedups when building large wheels.

Author: > We could split the paragraph here and also document why we might choose to override the defaults. I wrote about the reasoning in [#325 (comment)](https://github.com/mesonbuild/meson-python/pull/325#issuecomment-1463902837).

I checked it in more detail, and it's effectively the same as what I already have in this PR: _"because Meson defaults to values that are appropriate for development, while the main purpose of meson-python is to build release artifacts."_

Reviewer 2: > If we keep `meson install` I would much prefer if we would get rid of the `--prefix` and `--python.{plat,pure}lib` options and use the `$DESTDIR` environment variable pointing to somewhere in `/tmp` instead.

I don't think this is even a question, is it :rofl:? Those options and `DESTDIR` serve two different purposes. We only set those options for the heuristics, so they can be removed now.

Regarding `DESTDIR`, we already set it to a subdirectory of `working_dir`, which currently when using the PEP 517 hooks is always set by `tempfile.TemporaryDirectory` (in `with_temp_working_dir`), so isn't that true already?

Reviewer: The working dir is created inside the project directory, thus it is on the same filesystem as the project sources.

Reviewer 2: Ah, you're right :sweat_smile:. IIRC I did that to make things work properly for packagers by default, where access to `/tmp` might be a bit tricky. We can consider changing the default, though.

Author: I'll leave this comment thread open, because it seems like useful info. It doesn't apply to this PR directly though; it's discussion of a potential future design change. If that materializes, the docs can simply be updated.

``meson-python`` you have installed, look at the *User defined options* section
of the output during the configure stage of the build (e.g., by running
``python -m build --wheel``). This will look something like:

.. code-block:: text

    User defined options

>       Native files     : /home/username/code/project/.mesonpy-native-file.ini
>       debug            : false
>       optimization     : 2
>       prefix           : /home/username/mambaforge/envs/project-dev
>       python.platlibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
>       python.purelibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
>       b_ndebug         : if-release

Reviewer 2: possible improvement:

I'd group the options we need to set, with the ones we choose to override the default. Something like this:

```suggestion
      (we need override)
      prefix           : /home/username/mambaforge/envs/project-dev
      python.platlibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      python.purelibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      Native files     : /home/username/code/project/.mesonpy-native-file.ini
      (we chose to override)
      debug            : false
      optimization     : 2
      b_ndebug         : if-release
```

It might make sense to make the distinction, so that users know what they can and can't override.

Reviewer: This is the output of the `meson setup` command. Presenting an edited version of it that does not match what the uses sees would be confusing.

Author: Agreed, I just Ctrl-C'd this from actual build output (replacing my own user name with `username` and package name with `project`) and would prefer to leave it as is.

Author: rejecting the suggested change - resolving.

- meson-python uses a native file, written to the build dir and named
  ``mesonpy-native-file.ini``, in order to point Meson at the correct
  ``python`` interpreter to use (the same one for which ``meson-python`` was
  installed). This is necessary, because Meson may otherwise look for the first
  Python interpreter on the PATH (usually the same one, but not always the
  case). Users may use ``--native-file`` to pass a second native file to Meson;
  Meson will merge contents of both native file, so as long as the
  user-provided file does not try to pass a different path for the ``python``
  binary, this will work without a conflict.
- The ``prefix`` and ``platlibdir``/``purelibdir`` options also point Meson at
  that same interpreter and the environment in which it is installed.
- The ``debug``, ``optimization`` and ``b_ndebug`` options are overridden,
  because Meson defaults to values that are appropriate for development, while
  the main purpose of meson-python is to build release artifacts.

It is possible to override these defaults, either permanently in your project
or at build time. For example, to build a wheel with debug symbols, use:

> .. code-block:: bash

Reviewer 2: ```suggestion
.. code-block:: console
```

This makes it so that the $ is not selectable, excluding it from ctrl+a or the copy button from the sphinx extension.

Author: thanks, adopted

   $ python -m build -Csetup-args=-Ddebug=true

And to override all debug and optimization settings permanently, add this to
your ``pyproject.toml`` file:

.. code-block:: toml

   [tool.meson-python.args]

>    setup = ['-Ddebug=true', '-Doptimization=0', '-Db_ndebug=false']

Reviewer 2: Since #291 is not merged, wouldn't it be better to link to that page?

https://meson-python.readthedocs.io/en/latest/how-to-guides/meson-args.html

Author: Agreed, done.

> For more details on overriding build options at build time see
> :ref:`here <reference-config-settings>`, and in ``pyproject.toml`` see
> :ref:`here <reference-pyproject-settings>`.

Reviewer 2: nitpick

```suggestion
For more details on overriding build options at build time see the
:ref:`reference-config-settings` page, and in ``pyproject.toml`` see
the :ref:`reference-pyproject-settings` page.
```

Author: adopted, resolving
